### PR TITLE
Split locked / set states for home.  Tidy/Fix Plane's home-setting

### PR DESCRIPTION
**Pull request by peterbarker**
This fixes a problem in Plane where setting your home location (e.g. via a mavlink message) may reset your current waypoint.

`init_home` is removed in Plane on the basis that resetting your waypoint in the way that it does is not something you ever want to do apart from during the boot sequence (or, in the future, when home is first set).

The original bug can be reproduced in SITL by starting to fly the standard CMAC pattern then `long DO_SET_HOME 1`; you should see the vehicle skip from its current waypoint to the next.

### CODE REVIEW

File: libraries/AP_Common/AP_Common.h [hunk 1]
```diff
@@ -144,15 +144,6 @@ struct PACKED Location {
     int32_t lng;                                        ///< param 4 - Longitude * 10**7
 };
 
-/*
-  home states. Used to record if user has overridden home position.
-*/
-enum HomeState {
```
Comment (OXINARF) [inline libraries/AP_Common/AP_Common.h]: Great to remove this, but should be in its own commit instead of inside an *AP_AHRS* commit.

File: ArduPlane/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -791,7 +791,7 @@ void GCS_MAVLINK_Plane::handleMessage(mavlink_message_t* msg)
         case MAV_CMD_DO_SET_HOME: {
             result = MAV_RESULT_FAILED; // assume failure
             if (is_equal(packet.param1, 1.0f)) {
-                plane.init_home();
+                plane.set_home_persistently(AP::gps().location());
```
Comment (OXINARF) [inline ArduPlane/GCS_Mavlink.cpp]: Unrelated to this PR: doesn't this mean that in this case this will always respond with a failure?

File: ArduPlane/GCS_Plane.cpp [hunk 3]
```diff
@@ -11,3 +11,11 @@ void GCS_Plane::send_airspeed_calibration(const Vector3f &vg)
         }
     }
 }
+
+void GCS_Plane::send_set_home_statustext(const Location &loc)
```
Comment (OXINARF) [inline ArduPlane/GCS_Plane.cpp]: Calling @WickedShell to back me up on removing this ugly statustext 😄

File: libraries/AP_AHRS/AP_AHRS.h [hunk 4]
```diff
@@ -436,14 +436,18 @@ class AP_AHRS
         return _home;
     }
 
-    enum HomeState home_status(void) const {
-        return _home_status;
+    // functions to handle locking of home.  Some vehicles use this to
+    // allow GCS to lock in a home location.
+    void set_home_locked(bool locked) {
```
Comment (OXINARF) [inline libraries/AP_AHRS/AP_AHRS.h]: Any reason to not call this `lock_home(void)`? It seems we only want to lock, never unlock.

File: libraries/AP_Common/AP_Common.h (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Common/AP_Common.h]: Will fix.

File: ArduPlane/GCS_Mavlink.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduPlane/GCS_Mavlink.cpp]: Unrelated, but a bug.  I'll push a patch on top which resets `result`.

File: libraries/AP_AHRS/AP_AHRS.h (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS.h]: Eh.  Flexibility.  Doesn't hurt.

**Comment by peterbarker** [comment]
On Fri, 25 May 2018, WickedShell wrote:

> +    bool home_was_set = ahrs.home_is_set();
> 
> Why pull this out? You only use it once, and makes it a bit harder to trace what the authority is.

Because the return value of home_is_set will change after the next line...

> +    const bool home_was_set = ahrs.home_is_set();
> 
> Again, just move this to where you need the value.

:-)  Again, the same function call will have a different value next 
time...

> -                init_home();
> +                gcs().send_text(MAV_SEVERITY_INFO, "Init HOME");
> 
> If we're touching this status text (and not removing it), can we at least make this a bit more humane? "Initialized home" "Found home position" "Home
> position loaded" all come to mind. Anything that takes out "Init" and the caps lock on home gets my vote :)

That's a significant behavioural change which might break scripting, for 
example.  Not saying we shouldn't do it, just not my fight right now.

> +void GCS_Plane::send_set_home_statustext(const Location &loc)
> 
> KILL IT WITH FIRE! (PLEASE)

Hah!  That's definitely going to break scripts.  Not saying we shouldn't 
consider it, but... not my fight!

> In ArduPlane/GCS_Mavlink.cpp:
> 
> > @@ -1080,7 +1075,7 @@ void GCS_MAVLINK_Plane::handleMessage(mavlink_message_t* msg)
>              // param7 : altitude (absolute)
> 
> DO_SET_HOME is handled in both COMMAND_LONG and COMMAND_INT cases. Can we handle this same way MISSION_ITEM is where we simply convert it to a
> MISSION_ITEM_INT then pass it to a common handler? Would reduce the chance for differences here, and be less code.

Yes, that would be a good change.  Unrelated to this PR, however.

> +    // Save Home to EEPROM
> +    mission.write_home_to_storage();
> 
> Drop this, it's pointless. You cannot read home from storage, the entire concept of writing home to storage can be deleted. Seriously go ahead and delete
> write_home_to_storage() in this PR and you get credit! (And for bonus points and flash wear reasons, add a special case to write_cmd_to_storage() that
> refuses to write the first slot into storage.)

I'm happy to do more PRs, you know....

> +    // Save Home to EEPROM
> +    mission.write_home_to_storage();
> 
> Which leads us to the following conclusion: set_home_persistently() is a lie and can be deleted.

.... and *now* I start to wonder about actually doing that in this PR :-/

> +1 to making it a void function.

Did that.  For some values of "void function".

> +    bool _home_is_set :1;
> 
> Not that I actually object to this, but couldn't just checking that _home isn't lat/lng 0,0 sufficient? (Plane anyways can't navigate to 0,0 but I could
> be missing something about the no GPS use cases on other vehicles)

Magic values are EEEEVVVIIILLL...

I've pushed up patches.

**Comment by WickedShell** [comment]
@peterbarker you know e-mail/not continuing review thread responses are harder to follow right? :D

>> +void GCS_Plane::send_set_home_statustext(const Location &loc)
>>
>> KILL IT WITH FIRE! (PLEASE)

>Hah!  That's definitely going to break scripts.  Not saying we shouldn't consider it, but... not my fight!

Anyone who is dependent upon that statustext has written a bad scrip, there is a proper message for a reason. Who is the hold up on changing this?

> Magic values are EEEEVVVIIILLL...

I'm inclined to agree, I just wanted to ask :)
